dipy/align/imaffine.py

```diff
@@ -0,0 +1,1044 @@
+            self.moving_vals = np.array(self.moving_vals)
+            static_values = self.static_vals
+            moving_values = self.moving_vals
+        self.update_pdfs(static_values, moving_values)
+
+        # Compute the gradient of the joint PDF w.r.t. parameters
+        if update_gradient:
+            if self.samples is None:
+                # Compute the gradient of moving img. at physical points
+                # associated with the >>static image's grid<< cells
+                grid_to_world = M.dot(self.static_grid2world)
+                mgrad, inside = vf.gradient(self.moving,
+                                            self.moving_world2grid,
+                                            self.moving_spacing,
+                                            self.static.shape,
+                                            grid_to_world)
+                # Dense case: we just need to provide the grid-to-world
+                # transform to obtain the sampling points' world coordinates
+                # because we know that all grid cells must be processed
+                sampling_info = grid_to_world
+            else:
+                # Compute the gradient of moving at the sampling points
+                # which are already given in physical space coordinates
+                mgrad, inside = vf.sparse_gradient(self.moving,
+                                                   sp_to_moving,
+                                                   self.moving_spacing,
+                                                   self.samples)
+                # Sparse case: we need to provide the actual coordinates of
+                # all sampling points
+                sampling_info = self.samples[..., :self.dim]
+            self.update_gradient(params, self.transform, static_values,
+                                 moving_values, sampling_info, mgrad)
+
+        # Evaluate the mutual information and its gradient
+        # The results are in self.metric_val and self.metric_grad
+        # ready to be returned from 'distance' and 'gradient'
+        self.update_mi_metric(update_gradient)
+
+    def distance(self, params):
+        r""" Numeric value of the negative Mutual Information
+
+        We need to change the sign so we can use standard minimization
+        algorithms.
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        neg_mi : float
+            the negative mutual information of the input images after
+            transforming the moving image by the currently set transform
+            with `params` parameters
+        """
+        try:
+            self._update(params, False)
+        except:
```

Comment:
Unqualified `except:` statements very dangerous, as they can catch all sorts of things, like SyntaxErrors, that they are not meant to.   How about:

```
except AffineInversionError:
    return np.inf
```